clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java

```diff
@@ -693,11 +708,33 @@
 
     public static final Schema[] UPDATE_METADATA_REQUEST = new Schema[] {UPDATE_METADATA_REQUEST_V0, UPDATE_METADATA_REQUEST_V1};
     public static final Schema[] UPDATE_METADATA_RESPONSE = new Schema[] {UPDATE_METADATA_RESPONSE_V0, UPDATE_METADATA_RESPONSE_V1};
+    public static final short[] DEPRECATED_UPDATE_METADATA_REQUEST = new short[] {};
+
+    /* Protocol version api */
+    public static final Schema PROTOCOL_VERSION_REQUEST_V0 = new Schema();
+
+    public static final Schema SINGLE_PROTOCOL_VERSION_V0 = new Schema(
+        new Field("api_key", INT16, "Api key of protocol."),
+        new Field("api_name", STRING, "Api name of protocol."),
+        new Field("api_versions", new ArrayOf(INT16), "Supported api versions of protocol in order of preference."),
+        new Field("api_deprecated_versions", new ArrayOf(INT16), "Deprecated api versions of protocol.")
```

Comment:
How is a client supposed to use/handle deprecated versions? How much value does maintaining this  in code actually add?

I am thinking if we listed deprecated/removed protocol messages in the docs/release notes that should be good enough.